adcm_aio_client/core/requesters.py

```diff
@@ -22,20 +40,129 @@ def as_dict(self: Self) -> dict: ...
 
 
 class Requester(Protocol):
-    async def get(self: Self, path: str, query_params: dict) -> AsyncGenerator[RequesterResponse]: ...
+    async def get(self: Self, path: str, query_params: dict) -> RequesterResponse: ...
+
+    async def post(self: Self, path: str, data: dict) -> RequesterResponse: ...
+
+    async def patch(self: Self, path: str, data: dict) -> RequesterResponse: ...
+
+    async def delete(self: Self, path: str) -> RequesterResponse: ...
+
+
+@dataclass(slots=True, frozen=True)
+class Credentials:
+    username: str
+    password: str
+
+    def dict(self: Self) -> dict:
+        return asdict(self)
+
+    def __repr__(self: Self) -> str:
+        return f"{self.username}'s credentials"
+
+
+@dataclass(slots=True, frozen=True)
+class HTTPXRequesterResponse:
+    response: httpx.Response
+
+    def as_list(self: Self) -> list:
+        if not isinstance(data := self._prepare_json_response(), list):
+            message = f"Expected a list, got {type(data)}"
+            raise RequesterResponseError(message)
+
+        return data
+
+    def as_dict(self: Self) -> dict:
+        if not isinstance(data := self._prepare_json_response(), dict):
+            message = f"Expected a dict, got {type(data)}"
+            raise RequesterResponseError(message)
+
+        return data
 
-    async def post(self: Self, path: str, data: dict) -> AsyncGenerator[RequesterResponse]: ...
+    def _prepare_json_response(self: Self) -> Json:
+        try:
+            data = self.response.json()
```

Comment:
save data locally, no need to jsonify it every time (or is it cached on httpx level?)

Reply:
done